### PR TITLE
feat: support `no_std` for `reth-storage-errors`

### LINKED ISSUE #8459: Add no-std support for in reth-storage-errors
### Describe the feature

Ideally, we want to have a subset of crates with no_std supports (evm specifically)
we need to do some prep for this first by making some crates no_std first

## TODO

* add `#![cfg_attr(not(feature = "std"), no_std)]`
* Add to cargo.toml
```
default = ["std"]
std = []
``` 
* Fix errors by manually implementing std::error::Error instead of deriving and put it behind std feature

Best way to do this is `cargo expand` and modify the generated code

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by JackG-eth**
Description

have a subset of crates with no_std supports (evm specifically)
we need to do some prep for this first by making some crates no_std first

Motivation
closes #8459

### CODE REVIEW

File: crates/storage/errors/src/db.rs [hunk 1]
```diff
@@ -1,57 +1,97 @@
...
+use core::{
+    fmt::{Display, Result},
+    str::FromStr,
+};
+
+extern crate alloc;
+use alloc::boxed::Box;
+
+#[cfg(feature = "std")]
+use std::{error::Error, fmt};
```
Comment (mattsse) [inline crates/storage/errors/src/db.rs:10]: I prefer using the fully qualified path where necessary and not feature gate the `use` import

Comment (mattsse) [inline crates/storage/errors/src/db.rs:10]: fmt should be under core::fmt

File: crates/storage/errors/src/db.rs [hunk 2]
```diff
@@ -1,57 +1,97 @@
...
+            Self::Write(err) => Some(&**err),
+            _ => None,
+        }
+    }
+}
+
+impl Display for DatabaseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result {
+        match self {
+            Self::Open(info) => f.write_fmt(format_args!("failed to open the database: {0}", info)),
```
Comment (mattsse) [inline crates/storage/errors/src/db.rs:54]: I believe this can be replace with
`write!(f, "", args)`

File: crates/storage/errors/src/provider.rs [hunk 3]
```diff
@@ -154,17 +240,24 @@ pub struct RootMismatch {
     pub block_hash: BlockHash,
 }
 
+#[cfg(feature = "std")]
+impl std::error::Error for RootMismatch {}
```
Comment (mattsse) [inline crates/storage/errors/src/provider.rs:244]: like this

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/storage/errors/src/db.rs (hunk 1 above)
Comment (JackG-eth) [reply crates/storage/errors/src/db.rs:10]: Is this from a readability perspestive?

Comment (mattsse) [reply crates/storage/errors/src/db.rs]: yes and reduces cfgs

File: crates/storage/errors/src/db.rs [hunk 4]
```diff
@@ -1,57 +1,79 @@
-use std::{fmt::Display, str::FromStr};
-use thiserror::Error;
+use core::{
+    fmt::{Display, Formatter, Result},
+    str::FromStr,
+};
+
+extern crate alloc;
```
Comment (mattsse) [inline crates/storage/errors/src/db.rs:6]: this should be in lib.rs

File: crates/storage/errors/src/db.rs [hunk 5]
```diff
@@ -169,7 +211,7 @@ impl LogLevel {
 impl FromStr for LogLevel {
     type Err = String;
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
```
Comment (mattsse) [inline crates/storage/errors/src/db.rs:214]:
```suggestion
    fn from_str(s: &str) -> core::result::Result<Self, Self::Err> {
```

File: crates/storage/errors/src/provider.rs [hunk 6]
```diff
@@ -176,3 +266,19 @@ impl From<ConsistentViewError> for ProviderError {
...
+impl std::error::Error for ConsistentViewError {}
+
+impl Display for ConsistentViewError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            Self::Syncing { best_block } => {
+                f.write_fmt(format_args!("node is syncing. best block: {0:?}", best_block,))
+            }
+            Self::Inconsistent { tip } => {
+                f.write_fmt(format_args!("inconsistent database state: {0:?}", tip))
```
Comment (mattsse) [inline crates/storage/errors/src/provider.rs:280]: write!

**Review by mattsse — CHANGES_REQUESTED** [review]
last few nits

**Review by mattsse — CHANGES_REQUESTED** [review]
we don't have proper no-std tests in ci yet, so this currently passes, but fails with `cargo c --no-default-features`, mainly std related imports

**Comment by JackG-eth** [comment]
> we don't have proper no-std tests in ci yet, so this currently passes, but fails with `cargo c --no-default-features`, mainly std related imports

added necessary imports where needed, also added a couple more feature std flags where std libs are used.

**Review by mattsse — APPROVED** [review]
@DaniPopes I believe no-std-thiserror has a few limitations.

rolling this back once Error is available in no_std isn't a big deal so I'm fine with this, even if only temporary
wdyt?

**Pull request closed without merging**